core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1543,24 +1543,34 @@ class KafkaApis(val requestChannel: RequestChannel,
     var skippedMarkers = 0
     for (marker <- markers.asScala) {
       val producerId = marker.producerId
-      val (goodPartitions, partitionsWithIncorrectMessageFormat) = marker.partitions.asScala.partition { partition =>
+      val partitionsWithCorrectMessageFormat = new mutable.ListBuffer[TopicPartition]
+      val partitionsWithIncorrectMessageFormat = new mutable.ListBuffer[TopicPartition]
+      val unknownPartitions = new mutable.ListBuffer[TopicPartition]
+
+      marker.partitions.asScala.foreach { partition =>
         replicaManager.getMagic(partition) match {
-          case Some(magic) if magic >= RecordBatch.MAGIC_VALUE_V2 => true
-          case _ => false
+          case Some(magic) =>
+            if (magic < RecordBatch.MAGIC_VALUE_V2)
+              partitionsWithIncorrectMessageFormat.append(partition)
+            else
+              partitionsWithCorrectMessageFormat.append(partition)
+          case None =>
+            unknownPartitions.append(partition)
```

Comment:
Nit: you can use `+=` instead of `append` (e.g. `unknownPartitions += partition`).